plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+        }
+        return null
+    }
+
+    private fun IrExpression.getPureTypeValue(): IrExpression {
+        require(this is IrCall && isAtomicFactory()) { "Illegal initializer for the atomic property $this" }
+        return getValueArgument(0)!!
+    }
+
+    private fun IrExpression.buildPureTypeArrayConstructor() =
+        when (this) {
+            is IrConstructorCall -> {
+                require(isAtomicArrayConstructor()) { "Unsupported atomic array constructor $this" }
+                val arrayConstructorSymbol = type.getArrayConstructorSymbol { it.owner.valueParameters.size == 1 }
+                val size = getValueArgument(0)
+                IrConstructorCallImpl(
+                    UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                    irBuiltIns.unitType, arrayConstructorSymbol,
+                    0, 0, 1
+                ).apply {
+                    putValueArgument(0, size)
+                }
+            }
+            is IrCall -> {
+                require(isAtomicArrayFactory()) { "Unsupported atomic array factory function $this" }
+                val arrayFactorySymbol = referencePackageFunction("kotlin", "arrayOfNulls")
+                val arrayElementType = getTypeArgument(0)!!
+                val size = getValueArgument(0)
+                buildCall(
+                    target = arrayFactorySymbol,
+                    type = type,
+                    typeArguments = listOf(arrayElementType),
+                    valueArguments = listOf(size)
+                )
+            }
+            else -> error("Illegal type of atomic array initializer")
+        }
+
+    private fun IrCall.inlineAtomicFunction(atomicType: IrType, accessors: List<IrExpression>): IrCall {
+        val valueArguments = getValueArguments()
+        val functionName = getAtomicFunctionName()
+        val runtimeFunction = getRuntimeFunctionSymbol(functionName, atomicType)
+        return buildCall(
+            target = runtimeFunction,
+            type = type,
+            origin = IrStatementOrigin.INVOKE,
+            typeArguments = if (runtimeFunction.owner.typeParameters.size == 1) listOf(atomicType) else emptyList(),
+            valueArguments = valueArguments + accessors
+        )
+    }
+
+    private fun IrStatement.transformStatement(parentDeclaration: IrDeclaration, lambda: IrFunction? = null) =
+        when (this) {
+            is IrExpression -> transformAtomicFunctionCall(parentDeclaration, lambda)
+            is IrVariable -> {
+                apply { initializer = initializer?.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            else -> this
+        }
+
+    private fun IrExpression.transformAtomicFunctionCall(parentDeclaration: IrDeclaration, lambda: IrFunction? = null): IrExpression {
+        // erase unchecked cast to the Atomic* type
+        if (this is IrTypeOperatorCall && operator == IrTypeOperator.CAST && typeOperand.isAtomicValueType()) {
```

Comment:
I'd also check for `IrTypeOperator.IMPLICIT_CAST` since it covers the case of smart casts